### PR TITLE
Some numba backend fixes

### PR DESCRIPTION
**Pull request by aseyboldt**
- implementations for numba special functions
- numba impl fix for cumop
- Fix for numba svd
- Preserve names in graph replacements
- Remove broken AdvancdIndexing numba op
- Deal with negative axis args
- numba impl for checkandraise

### CODE REVIEW

**Comment by twiecki** [comment]
@aseyboldt Are you porting over a PR from aesara? If so, you can use the `bin/downstream_pr.sh` script to auto-rename.

**Comment by aseyboldt** [comment]
Yes, I just wanted to make sure we don't duplicate that work

File: aesara/graph/fg.py [hunk 1]
```diff
@@ -484,7 +484,9 @@ def replace(
                 f"rewriting: rewrite {reason} replaces {var} of {var.owner} with {new_var} of {new_var.owner}"
             )
 
+        name = new_var.name
         new_var = var.type.filter_variable(new_var, allow_convert=True)
+        new_var.name = name
```
Comment (ricardoV94) [inline aesara/graph/fg.py:489]: Do we need this in general? In PyMC I added a fix here: https://github.com/pymc-devs/pymc/pull/6336/commits/dc3323a4fb9db09863b059160fcfb31ed4b0fc1c

I am not sure we want to call `SpecifyShape(x)`, `x.name` in general. We might need to, but what was the issue you found?

File: aesara/link/numba/dispatch/elemwise.py [hunk 2]
```diff
@@ -164,6 +164,18 @@ def create_vectorize_func(
...
+def normalize_axis(axis, ndim):
+    if axis is None:
+        return axis
+
+    if axis < 0:
+        axis = ndim + axis
+
+    if axis < 0 or axis >= ndim:
+        raise np.AxisError(ndim=ndim, axis=axis)
+    return axis
```
Comment (ricardoV94) [inline aesara/link/numba/dispatch/elemwise.py:176]: Can you just use https://github.com/numpy/numpy/blob/7f0f045625022c3f816911cd80f8635ac2a36f21/numpy/core/numeric.py#L1348

File: aesara/graph/fg.py (hunk 1 above)
Comment (ricardoV94) [reply aesara/graph/fg.py:489]: Also in some rewrites like MergeOptimizer, you may be erasing the name if the "reference variable" name is `None` or something else

File: aesara/link/numba/dispatch/basic.py [hunk 3]
```diff
@@ -48,10 +48,14 @@
 
 def numba_njit(*args, **kwargs):
 
+    kwargs = kwargs.copy()
+    if "cache" not in kwargs:
+        kwargs["cache"] = config.numba__cache
```
Comment (ferrine) [inline aesara/link/numba/dispatch/basic.py:53]:
```suggestion
    kwargs.setdefault("cache", config.numba__cache)
```

File: aesara/graph/fg.py (hunk 1 above)
Comment (aseyboldt) [reply aesara/graph/fg.py:489]: I think that was to fix the same issue you fixed here: https://github.com/pymc-devs/pymc/pull/6336/commits/3118cc646c6b7a4324184274fdbf545a5a32773b
I'll remove the commit in the rebase.

File: aesara/link/numba/dispatch/elemwise.py [hunk 4]
```diff
@@ -218,6 +230,8 @@ def careduce_axis(x):
 
     """
 
+    axis = normalize_axis(axis, ndim)
```
[the remark this answers is not in the document]
Comment (aseyboldt) [reply aesara/link/numba/dispatch/elemwise.py:233]: I saw some issues in some ops due to this (I think cumop), but thought I'll just fix it in different place as well. I think we can keep it ever twice?

File: aesara/graph/fg.py (hunk 1 above)
Comment (aseyboldt) [reply aesara/graph/fg.py]: done

File: aesara/link/numba/dispatch/elemwise.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (aseyboldt) [reply aesara/link/numba/dispatch/elemwise.py]: Does that actually do what we want? It doesn't work if axis is None and raises a TypeError instead. And the `max(1, ndim)` also seems a bit suspicious: Shouldn't it throw an error if I specify a non-None axis for a rank 0 array? (as `normalize_axis_index` does)

[the remark this answers is not in the document]
Comment (ricardoV94) [reply aesara/link/numba/dispatch/elemwise.py]: That question was brought up in the original PR, the answer in the comment above holds I think: https://github.com/pymc-devs/pytensor/blob/b7a2d30830da8b201ab47879699a5a7553193fdc/pytensor/tensor/elemwise.py#L1315

[the remark this answers is not in the document]
Comment (ricardoV94) [reply aesara/link/numba/dispatch/elemwise.py]: Axis None is handled above:  https://github.com/pymc-devs/pytensor/blob/b7a2d30830da8b201ab47879699a5a7553193fdc/pytensor/tensor/elemwise.py#L1311-L1312

[the remark this answers is not in the document]
Comment (ricardoV94) [reply aesara/link/numba/dispatch/elemwise.py]: Or do you mean something else?

[the remark this answers is not in the document]
Comment (ricardoV94) [reply aesara/link/numba/dispatch/elemwise.py]: But to answer your question, sure we can have duplicated logic. Ideally `make_node` would normalize axis though so we don't have to worry about it during rewrites and stuff like that (e.g., `MergeOptimizer`).

[the remark this answers is not in the document]
Comment (aseyboldt) [reply aesara/link/numba/dispatch/elemwise.py]: Mostly this additional check is in helper functions for generating reductions etc, and it is not obvious which ops might use this in the future. I think it makes sense to have the axis normalization in there as well, to make sure we don't run into issues with that in the future, when we add new op implementations.

File: aesara/link/numba/dispatch/elemwise.py (hunk 2 above)
Comment (ricardoV94) [reply aesara/link/numba/dispatch/elemwise.py]: Dod you decide not to use these?

**Review by ricardoV94 — APPROVED** [review]
LGTM, just the nitpick not to reinvent the numpy axis helpers.

File: pytensor/link/numba/dispatch/cython_support.py [hunk 5]
```diff
@@ -0,0 +1,211 @@
...
+    "double": np.double,
+    "long double": np.longdouble,
+    "float complex": np.csingle,
+    "double complex": np.cdouble,
+}
+
+
+@dataclass
+class Signature:
+    res_dtype: DTypeLike
```
Comment (ferrine) [inline pytensor/link/numba/dispatch/cython_support.py:34]: is it single output signature?

File: aesara/link/numba/dispatch/elemwise.py (hunk 2 above)
Comment (aseyboldt) [reply aesara/link/numba/dispatch/elemwise.py]: sorry, done

File: pytensor/link/numba/dispatch/cython_support.py (hunk 5 above)
Comment (aseyboldt) [reply pytensor/link/numba/dispatch/cython_support.py:34]: Yes, this currently only supports scalar return values. This is all we need for the scipy.special functions, but if we need more later (maybe for some linalg stuff?) we might have to expand it a bit.

**Review by ricardoV94 — APPROVED** [review]

**Pull request merged**